bugsnag-android-core/src/main/java/com/bugsnag/android/FileStore.kt

```diff
@@ -115,23 +112,21 @@ internal abstract class FileStore(
         // Limit number of saved payloads to prevent disk space issues
         if (isStorageDirValid(storageDir)) {
             val listFiles = storageDir.listFiles() ?: return
-            val files: ArrayList<File> = arrayListOf(*listFiles)
-            if (files.size >= maxStoreCount) {
-                // Sort files then delete the first one (oldest timestamp)
-                Collections.sort(files, comparator)
-                var k = 0
-                while (k < files.size && files.size >= maxStoreCount) {
-                    val oldestFile = files[k]
-                    if (!queuedFiles.contains(oldestFile)) {
-                        logger.w(
-                            "Discarding oldest error as stored " +
-                                "error limit reached: '" + oldestFile.path + '\''
-                        )
-                        deleteStoredFiles(setOf(oldestFile))
-                        files.removeAt(k)
-                        k--
-                    }
-                    k++
+            if (listFiles.size < maxStoreCount) return
+            val sortedListFiles = listFiles.sortedBy { it.lastModified() }
+            // Number of files to discard takes into account that a new file may need to be written
+            val numberToDiscard = listFiles.size - maxStoreCount + 1
+            var discardedCount = 0
+            for (file in sortedListFiles) {
+                if (discardedCount == numberToDiscard) {
+                    return
+                } else if (!queuedFiles.contains(file)) {
+                    logger.w(
+                        "Discarding oldest error as stored error limit reached: '" +
+                            file.path + '\''
+                    )
+                    deleteStoredFiles(setOf(file))
+                    discardedCount++
                 }
             }
```

Comment:
I suspect we can further simplify this logic using a Sequence:
```suggestion
            deleteStoredFiles(
                listFiles
                    .asSequence()
                    // only discard files that are not currently in the queue
                    .filter { it !in queuedFiles }
                    // discard the oldest files first
                    .sortedBy { it.lastModified() }
                    // take the 'numberToDiscard' oldest files off the top of the sequence
                    .take(numberToDiscard)
                    // log each file that will be discarded
                    .onEach {
                        logger.w(
                            "Discarding oldest error as stored error limit reached: '${it.path}'"
                        )
                    }
                    .toSet()
            )
```